### ISSUE TITLE
Connector Mappings are not handled correctly by Variable tooling

### ISSUE THREAD

**Issue by marstamm** [camunda/camunda-modeler#3524]
**Describe the bug**

When using connectors, the editor tooling offers no or wrong variable support.

**How to reproduce**

__No support__

1. use a REST connector
2. Add a result Variable or Result Expression
3. Try to use it in a Service Task
4. No Variables are suggested

![image](https://user-images.githubusercontent.com/21984219/228492712-f53c52d5-c3fd-4061-8f5f-0e00c0da7d52.png) ![image](https://user-images.githubusercontent.com/21984219/228492836-1bad281b-a48f-442f-8da2-891711360eef.png)

__Wrong support__

In Connector, output mappings, only explicitly linked data is available.
1. Have a process with variables in process scope (e.g. by creating an output mapping on a task)
2. Create a `Result Expression` in a Connector
3. The global variable is suggested

![image](https://user-images.githubusercontent.com/21984219/228497430-2763ba2f-7e60-422e-863b-cb773596dfe0.png)

**Expected Behavior**
The Variable integration supports connectors

**Additional Context**
Connectors use `zeebe:Taskheader` to define mappings and create the mappings in the Task worker. This means that the variables are created outside of bpmn/zeebe engine context.
I [confirmed with the connectors team](https://camunda.slack.com/archives/C02JLRNQQ05/p1680075717859209) that the same headers are used for mappings by all connectors.

The behavior as discussed with @nikku is the following:
- `resultVariable` will add the response to the defined variable and return it
- `resultExpression` will be evaluated and **ALL KEYS** are then mapped to an individual process variable
- If both are set, both are evaluated and created
- If none are set, no process variable will be set. `response` will be unavailable in the process as well. 

**Possible Solutions**

__Treat headers as output mappings__
This special behavior puts us in a difficult position for modeling. We could treat the headers as output mappings, however 
> If none are set, no process variable will be set. `response` will be unavailable in the process as well

puts us in a difficult decision. When no Task header is set and an element template is used, we can't determine if additional variables should
 - be mapped to the process ("non-connector template")
 - not be passed (connector template)

Additionally, unlinking a template would influence the data handling during modeling but has no impact on the execution. **Logic hidden in the task worker can not be reliably handled in editor tooling**.

Therefore, I propose we discover possible solutions on the connectors side as well and evaluate https://github.com/camunda/connector-sdk/issues/312#issuecomment-1405192073 again. Making the output mappings an engine concern would allow us to build upon shared behavior independent of Task worker implementations.

---

Depends on https://github.com/camunda/camunda-bpmn-js/issues/265
Referenced in [this discussion](https://camunda.slack.com/archives/GP70M0J6M/p1692000608571909).

**Comment by nikku** [comment]
As an immediate remedy I propose the following:

* Only provide variable completion in places where we know about scoping, i.e. fields backed by IO mappings

**Comment by nikku** [comment]
@marstamm I created a follow up issue for this. https://github.com/camunda/camunda-modeler/issues/3525.

**Comment by nikku** [comment]
We were able to mitigate wrong suggestions (https://github.com/camunda/camunda-modeler/issues/3525). So at this point you:

(1) get completion in input and output bindings (as expected)
(2) won't get completions in C8 connectors with custom result mapping

This is a safe and sane foundation to further explore.

**Comment by marstamm** [comment]
> @marstamm Does this depend on https://github.com/camunda/camunda-bpmn-js/issues/265 or duplicate it? If so, lets clean this up, or mark as Depends on #.....

I think it depends on it (now marked as such). It is a sub-problem of the overarching issue of connector support

**Comment by marstamm** [comment]
## Proposal for intermediate solution

As the full connector behavior can not be replicated in the Frontend, here is my proposal for the way forward so we are not blocked while further developing the data providers:

> Handle `resultVariable` and `resultExpression` the same as `OutputMappings`

This would allow us to have correct behavior in most cases. The only exceptions are:

- Connectors that don't have a result
- ElementTemplates/Elements using these `zeebeHeaders` outside of the connector runtime

This allows us to make progress in the connector data story while we don't have mapping handled in engine behavior. For backwards-compatibility, we would probably need such a solution at some point anyway.

**Comment by nikku** [comment]
What compromises are we willing to take here? We'd still need to solve the scoping issue so we don't delude our users, for example:

> Handle resultVariable and resultExpression the same as OutputMappings

Here you only have access to variables mapped as inputs, and connector results, not process variables.

> However, missing Headers would still result in wrong data being suggested while modeling

What contract would we need to end up in a better behavior? Shall the header be mandatory for connectors? Or could we simply assume that for elements backed by a template output is always explicit? While that is not true it would result in a SAFER behavior (rather show less suggestions than more).

**Comment by marstamm** [comment]
> Here you only have access to variables mapped as inputs, and connector results, not process variables.

We could not provide suggestions on the mapping itself (or it would require some hacky special-cases). This would at least add the defined variables to the process and make discoverability easier/more predictable. This would only solve part of the issue ("No support" for defined variables).

> What contract would we need to end up in a better behavior?

Explicit Headers would not work with existing connectors. I propose, to get a consistent behavior, we tie the behavior not to the element template, but the data source. So data provided by the `additionalData` Tab would always need to be backed by an output mapping, independent of where it is used.
That would mean that data provided by JSON examples would never leak into the parent scope and always need explicit mappings. This would only suggest defined variables (safe), promote explicit mappings and work when element templates are unlinked

**Comment by nikku** [comment]
> This would only suggest defined variables (safe), promote explicit mappings and work when element templates are unlinked

If we can end up with such a solution I'm all in for it :tada: 

> Also, I think this can be achieved in the individual variable providers, so this behavior can be changed depending on the provider. 

Yea, that would be my suggestion, too.

**Comment by nikku** [comment]
[This thread](https://camunda.slack.com/archives/GP70M0J6M/p1692000608571909) added additional evidence that we may want to move forward with this topic.

**Comment by nikku** [comment]
Scheduled alignment session with @sbuettner.

**Comment by marstamm** [comment]
As a lot of the originally discovered problems were solved, I created a follow-up issue to focus on the output mapping problem only: https://github.com/camunda/camunda-modeler/issues/3811

**Comment by nikku** [comment]
Marking this as `M68`. Let's brag about improved connector variables support!
